File: plugins/expo-autofill-plugin/ios-template/PearPassAutofillExtension/BareHelper.m
#import "BareHelper.h"

@interface BareHelper ()
@property (nonatomic, strong) BareWorklet *worklet;
@property (nonatomic, strong) BareIPC *ipc;
@property (nonatomic, strong) NSString *bundleName;
@property (nonatomic, strong) NSString *bundleType;
@property (nonatomic, assign) NSUInteger memoryLimit;
@property (nonatomic, assign) BOOL isRunning;
@end

@implementation BareHelper

- (instancetype)initWithBundleName:(NSString *)bundleName 
                         bundleType:(NSString *)bundleType 
                        memoryLimit:(NSUInteger)memoryLimitInMB {
    self = [super init];
    if (self) {
        _bundleName = bundleName;
        _bundleType = bundleType;
        _memoryLimit = memoryLimitInMB * 1024 * 1024; // Convert MB to bytes
        _isRunning = NO;
        NSLog(@"BareHelper: Initialized with bundle %@.%@ and memory limit: %lu bytes", 
              _bundleName, _bundleType, (unsigned long)_memoryLimit);
    }
    return self;
}

- (BOOL)startWorklet {
    if (self.isRunning) {
        NSLog(@"BareHelper: Worklet is already running");
        return NO;
    }
    
    NSLog(@"BareHelper: Starting worklet with bundle: %@.%@", self.bundleName, self.bundleType);
    
    // Create worklet configuration
    BareWorkletConfiguration *config = [BareWorkletConfiguration defaultWorkletConfiguration];
    config.memoryLimit = self.memoryLimit;
    
    // Create worklet
    self.worklet = [[BareWorklet alloc] initWithConfiguration:config];
    if (!self.worklet) {
        NSLog(@"BareHelper: Failed to create worklet");
        return NO;
    }
    
    NSLog(@"BareHelper: Worklet created successfully");
    
    // Start worklet with bundle
    @try {
        [self.worklet start:self.bundleName ofType:self.bundleType arguments:@[]];
        NSLog(@"BareHelper: Worklet started successfully with bundle");
    } @catch (NSException *exception) {
        NSLog(@"BareHelper: Failed to start worklet: %@", exception.reason);
        self.worklet = nil;
        return NO;
    }
    
    // Initialize IPC
    self.ipc = [[BareIPC alloc] initWithWorklet:self.worklet];
    if (!self.ipc) {
        NSLog(@"BareHelper: Failed to create IPC");
        [self.worklet terminate];
        self.worklet = nil;
        return NO;
    }
    
    NSLog(@"BareHelper: IPC created successfully");
    self.isRunning = YES;
    
    return YES;
}

- (void)write:(NSData *)data completion:(BareHelperWriteCompletion)completion {
    if (!self.isRunning || !self.ipc) {
        NSLog(@"BareHelper: Cannot write - worklet not running or IPC not available");
        if (completion) {
            NSError *error = [NSError errorWithDomain:@"BareHelper" 
                                                 code:1001 
                                             userInfo:@{NSLocalizedDescriptionKey: @"Worklet not running"}];
            completion(error);
        }
        return;
    }
    
    NSLog(@"BareHelper: Writing %lu bytes", (unsigned long)data.length);
    [self.ipc write:data completion:completion];
}

- (void)read:(BareHelperReadCompletion)completion {
    if (!self.isRunning || !self.ipc) {
        NSLog(@"BareHelper: Cannot read - worklet not running or IPC not available");
        if (completion) {
            NSError *error = [NSError errorWithDomain:@"BareHelper" 
                                                 code:1002 
                                             userInfo:@{NSLocalizedDescriptionKey: @"Worklet not running"}];
            completion(nil, error);
        }
        return;
    }
    
    NSLog(@"BareHelper: Reading data");
    [self.ipc read:completion];
}

- (void)send:(NSString *)message completion:(BareHelperSendCompletion)completion {
    if (!self.isRunning || !self.ipc) {
        NSLog(@"BareHelper: Cannot send - worklet not running or IPC not available");
        if (completion) {
            NSError *error = [NSError errorWithDomain:@"BareHelper" 
                                                 code:1003 
                                             userInfo:@{NSLocalizedDescriptionKey: @"Worklet not running"}];
            completion(nil, error);
        }
        return;
    }
    
    // Convert string to data
    NSData *messageData = [message dataUsingEncoding:NSUTF8StringEncoding];
    if (!messageData) {
        NSLog(@"BareHelper: Failed to convert message to data");
        if (completion) {
            NSError *error = [NSError errorWithDomain:@"BareHelper" 
                                                 code:1004 
                                             userInfo:@{NSLocalizedDescriptionKey: @"Failed to encode message"}];
            completion(nil, error);
        }
        return;
    }
    
    NSLog(@"BareHelper: Sending message: %@", message);
    
    // Use weak/strong pattern to avoid retain cycles and ensure IPC remains valid
    __weak typeof(self) weakSelf = self;
    
    // Write the message
    [self.ipc write:messageData completion:^(NSError * _Nullable writeError) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (!strongSelf) {
            NSLog(@"BareHelper: Self was deallocated during write");
            return;
        }
        
        if (writeError) {
            NSLog(@"BareHelper: Write failed: %@", writeError.localizedDescription);
            if (completion) {
                completion(nil, writeError);
            }
            return;
        }
        
        NSLog(@"BareHelper: Message sent, waiting for reply");
        
        // Check if IPC is still valid before reading
        if (!strongSelf.ipc || !strongSelf.isRunning) {
            NSLog(@"BareHelper: IPC no longer valid after write");
            if (completion) {
                NSError *error = [NSError errorWithDomain:@"BareHelper" 
                                                     code:1007 
                                                 userInfo:@{NSLocalizedDescriptionKey: @"IPC connection lost"}];
                completion(nil, error);
            }
            return;
        }
        
        // Read the reply
        [strongSelf.ipc read:^(NSData * _Nullable replyData, NSError * _Nullable readError) {
            if (readError) {
                NSLog(@"BareHelper: Read failed: %@", readError.localizedDescription);
                if (completion) {
                    completion(nil, readError);
                }
                return;
            }
            
            if (!replyData) {
                NSLog(@"BareHelper: No reply data received");
                if (completion) {
                    NSError *error = [NSError errorWithDomain:@"BareHelper" 
                                                         code:1005 
                                                     userInfo:@{NSLocalizedDescriptionKey: @"No reply data received"}];
                    completion(nil, error);
                }
                return;
            }
            
            // Convert reply data to string
            NSString *reply = [[NSString alloc] initWithData:replyData encoding:NSUTF8StringEncoding];
            if (!reply) {
                NSLog(@"BareHelper: Failed to decode reply data (data length: %lu)", (unsigned long)replyData.length);
                if (completion) {
                    NSError *error = [NSError errorWithDomain:@"BareHelper" 
                                                         code:1006 
                                                     userInfo:@{NSLocalizedDescriptionKey: @"Failed to decode reply"}];
                    completion(nil, error);
                }
                return;
            }
            
            NSLog(@"BareHelper: Received reply: %@", reply);
            if (completion) {
                completion(reply, nil);
            }
        }];
    }];
}

- (void)shutdown {
    NSLog(@"BareHelper: Shutting down");
    
    if (self.ipc) {
        [self.ipc close];
        self.ipc = nil;
    }
    
    if (self.worklet) {
        [self.worklet terminate];
        self.worklet = nil;
    }
    
    self.isRunning = NO;
    NSLog(@"BareHelper: Shutdown complete");
}

- (void)dealloc {
    [self shutdown];
    NSLog(@"BareHelper: Deallocated");
}

@end

File: plugins/expo-autofill-plugin/android-template/java/autofill/data/BareHelper.java
package com.pears.pass.autofill.data;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import com.pears.pass.autofill.utils.SecureLog;
import java.io.IOException;
import java.io.InputStream;
import java.io.ByteArrayOutputStream;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import to.holepunch.bare.kit.IPC;
import to.holepunch.bare.kit.Worklet;

public class BareHelper {
    private static final String TAG = "BareHelper";

    private final Context context;
    private final String bundleName;
    private final String bundleType;
    private final int memoryLimit;
    private Worklet worklet;
    private IPC ipc;
    private boolean isRunning = false;

    public interface WriteCallback {
        void onComplete(Throwable error);
    }

    public interface ReadCallback {
        void onData(byte[] data, Throwable error);
    }

    public interface SendCallback {
        void onResponse(String reply, Throwable error);
    }

    public BareHelper(Context context, String bundleName, String bundleType, int memoryLimitInMB) {
        this.context = context;
        this.bundleName = bundleName;
        this.bundleType = bundleType;
        this.memoryLimit = memoryLimitInMB * 1024 * 1024; // Convert MB to bytes
        SecureLog.d(TAG, "Initialized with bundle " + bundleName + "." + bundleType + " and memory limit: " + memoryLimit + " bytes");
    }

    public boolean startWorklet() {
        if (isRunning) {
            SecureLog.d(TAG, "Worklet is already running");
            return false;
        }

        SecureLog.d(TAG, "Starting worklet with bundle: " + bundleName + "." + bundleType);

        try {
            // Create worklet configuration
            Worklet.Options config = new Worklet.Options();
            config.memoryLimit = memoryLimit;

            // Create worklet
            worklet = new Worklet(config);

            SecureLog.d(TAG, "Worklet created successfully");

            // Start worklet with bundle
            // The bundle file is located in the assets directory
            String bundleFileName = bundleName + "." + bundleType;
            InputStream bundleStream = context.getAssets().open(bundleFileName);
            worklet.start("/" + bundleFileName, bundleStream, new String[]{});

            SecureLog.d(TAG, "Worklet started successfully with bundle");

            // Initialize IPC
            ipc = new IPC(worklet);

            SecureLog.d(TAG, "IPC created successfully - IPC: " + ipc + ", Worklet: " + worklet);

            isRunning = true;

            return true;
        } catch (Exception e) {
            SecureLog.e(TAG, "Failed to start worklet", e);
            // Don't call worklet.terminate() here - if start() failed,
            // the native instance may not be properly initialized and
            // terminate() will crash with SIGSEGV (null pointer dereference)
            worklet = null;
            return false;
        }
    }

    public void write(byte[] data, WriteCallback callback) {
        if (!isRunning || ipc == null) {
            SecureLog.e(TAG, "Cannot write - worklet not running or IPC not available");
            if (callback != null) {
                callback.onComplete(new Exception("Worklet not running"));
            }
            return;
        }

        SecureLog.d(TAG, "Writing " + data.length + " bytes");

        try {
            ByteBuffer buffer = ByteBuffer.allocateDirect(data.length);
            buffer.put(data);
            buffer.flip();

            ipc.write(buffer, (error) -> {
                if (callback != null) {
                    callback.onComplete(error);
                }

                SecureLog.d(TAG, "Writing done!");
            });
        } catch (Exception e) {
            SecureLog.e(TAG, "Error during write operation", e);
            if (callback != null) {
                callback.onComplete(e);
            }
        }
    }

    public void read(ReadCallback callback) {
        if (!isRunning || ipc == null) {
            SecureLog.e(TAG, "Cannot read - worklet not running or IPC not available");
            if (callback != null) {
                callback.onData(null, new Exception("Worklet not running"));
            }
            return;
        }

        if (worklet == null) {
            SecureLog.e(TAG, "Cannot read - worklet is null");
            if (callback != null) {
                callback.onData(null, new Exception("Worklet is null"));
            }
            return;
        }

        SecureLog.d(TAG, "Reading data - IPC: " + ipc + ", Worklet: " + worklet);

        try {
            // The crash happens here - let's check if the IPC might have an issue with its internal state
            // The SIGSEGV at 0x18 suggests accessing a field at offset 0x18 of a null object

            // Let's try to validate the IPC state before calling read
            SecureLog.d(TAG, "About to call ipc.read() - validating state");

            ipc.read((data, error) -> {
                SecureLog.d(TAG, "IPC read callback invoked");
                try {
                    if (error != null) {
                        SecureLog.e(TAG, "IPC read error: " + error.getMessage());
                        if (callback != null) {
                            callback.onData(null, error);
                        }
                    } else if (data != null && data.remaining() > 0) {
                        byte[] bytes = new byte[data.remaining()];
                        data.get(bytes);
                        SecureLog.d(TAG, "Read " + bytes.length + " bytes");
                        if (callback != null) {
                            callback.onData(bytes, null);
                        }
                    } else {
                        SecureLog.w(TAG, "No data available to read");
                        if (callback != null) {
                            callback.onData(null, new Exception("No data received"));
                        }
                    }
                } catch (Exception e) {
                    SecureLog.e(TAG, "Error processing read data", e);
                    if (callback != null) {
                        callback.onData(null, e);
                    }
                }
            });
        } catch (Exception e) {
            SecureLog.e(TAG, "Error initiating read: " + e.getMessage(), e);
            if (callback != null) {
                callback.onData(null, e);
            }
        }
    }

    public void send(String message, SendCallback callback) {
        if (!isRunning || ipc == null) {
            SecureLog.e(TAG, "Cannot send - worklet not running or IPC not available");
            if (callback != null) {
                callback.onResponse(null, new Exception("Worklet not running"));
            }
            return;
        }

        // Convert string to bytes
        byte[] messageData;
        try {
            messageData = message.getBytes(StandardCharsets.UTF_8);
        } catch (Exception e) {
            SecureLog.e(TAG, "Failed to encode message", e);
            if (callback != null) {
                callback.onResponse(null, e);
            }
            return;
        }

        SecureLog.d(TAG, "Sending message: " + message);

        try {
            ByteBuffer writeBuffer = ByteBuffer.wrap(messageData);
            
            ipc.write(writeBuffer, (writeException) -> {
                if (writeException != null) {
                    SecureLog.e(TAG, "Write failed", writeException);
                    if (callback != null) {
                        callback.onResponse(null, writeException);
                    }
                    return;
                }

                SecureLog.d(TAG, "Message written successfully, now reading response");

                // Read the response using length-prefixed framing.
                // The worklet sends: [4-byte Uint32LE totalLength][data...]
                // Created via FramedStream on the JS side — read the 4-byte
                // length prefix first, then accumulate payload bytes.
                readLengthPrefix(new ByteArrayOutputStream(), callback);
            });

        } catch (Exception e) {
            SecureLog.e(TAG, "Error in send operation", e);
            if (callback != null) {
                callback.onResponse(null, e);
            }
        }
    }

    /** Maximum bytes we'll accumulate per response (10MB). */
    private static final int MAX_RESPONSE_BYTES = 10 * 1024 * 1024;

    /**
     * Read a 4-byte length prefix from IPC, then read that many payload bytes
     * across potentially multiple IPC reads.
     */
    private void readLengthPrefix(ByteArrayOutputStream accumulator, SendCallback callback) {
        ipc.read((replyData, readException) -> {
            if (readException != null) {
                SecureLog.e(TAG, "Read failed", readException);
                if (callback != null) {
                    callback.onResponse(null, readException);
                }
                return;
            }

            if (replyData == null || replyData.remaining() == 0) {
                SecureLog.w(TAG, "No reply data received");
                if (callback != null) {
                    callback.onResponse(null, new Exception("No reply data received"));
                }
                return;
            }

            // Read this chunk
            byte[] chunk = new byte[replyData.remaining()];
            replyData.get(chunk);
            accumulator.write(chunk, 0, chunk.length);
            byte[] accumulated = accumulator.toByteArray();

            // Accumulate until we have 4 bytes for the length prefix
            if (accumulated.length < 4) {
                readLengthPrefix(accumulator, callback);
                return;
            }
            int totalLen = ((accumulated[0] & 0xFF)) |
                          ((accumulated[1] & 0xFF) << 8) |
                          ((accumulated[2] & 0xFF) << 16) |
                          ((accumulated[3] & 0xFF) << 24);
            if (totalLen <= 0 || totalLen > MAX_RESPONSE_BYTES) {
                SecureLog.e(TAG, "Invalid response length: " + totalLen);
                if (callback != null) {
                    callback.onResponse(null, new Exception("Invalid response length: " + totalLen));
                }
                return;
            }
            SecureLog.d(TAG, "Response length prefix: " + totalLen + " bytes");
            // Reset accumulator and proceed to read payload
            ByteArrayOutputStream payloadAccum = new ByteArrayOutputStream(totalLen);
            // Move any excess bytes beyond the 4-byte prefix into the payload
            if (accumulated.length > 4) {
                payloadAccum.write(accumulated, 4, accumulated.length - 4);
            }
            readPayload(payloadAccum, totalLen, callback);
        });
    }

    private void readPayload(ByteArrayOutputStream accumulator, int totalLen, SendCallback callback) {
        // If we already have enough bytes, decode and return
        if (accumulator.size() >= totalLen) {
            byte[] data = accumulator.toByteArray();
            String reply = new String(data, 0, totalLen, StandardCharsets.UTF_8);
            SecureLog.d(TAG, "Complete response received (" + totalLen + " bytes)");
            if (callback != null) {
                callback.onResponse(reply, null);
            }
            return;
        }

        // Need more data — read next chunk
        ipc.read((replyData, readException) -> {
            if (readException != null) {
                SecureLog.e(TAG, "Read failed during payload", readException);
                if (callback != null) {
                    callback.onResponse(null, readException);
                }
                return;
            }

            if (replyData == null || replyData.remaining() == 0) {
                // No more data but we haven't reached totalLen — truncated
                byte[] partial = accumulator.toByteArray();
                SecureLog.e(TAG, "Response truncated: got " + accumulator.size() + "/" + totalLen + " bytes");
                if (callback != null) {
                    callback.onResponse(null, new Exception("Response truncated: " + accumulator.size() + "/" + totalLen + " bytes"));
                }
                return;
            }

            byte[] chunk = new byte[replyData.remaining()];
            replyData.get(chunk);
            accumulator.write(chunk, 0, chunk.length);

            readPayload(accumulator, totalLen, callback);
        });
    }



    public void shutdown() {
        SecureLog.d(TAG, "Shutting down");

        if (ipc != null) {
            ipc.close();
            ipc = null;
        }

        if (worklet != null) {
            worklet.terminate();
            worklet = null;
        }

        isRunning = false;
        SecureLog.d(TAG, "Shutdown complete");
    }

    public boolean isRunning() {
        return isRunning;
    }

    @Override
    protected void finalize() throws Throwable {
        shutdown();
        SecureLog.d(TAG, "Finalized");
        super.finalize();
    }
}